s3fs/core.py

```diff
@@ -160,45 +161,58 @@ def __init__(self, anon=False, key=None, secret=None, token=None,
             secret = password
 
         self.anon = anon
-        self.session = None
-        self.passed_in_session = session
-        if self.passed_in_session:
-            self.session = self.passed_in_session
         self.key = key
         self.secret = secret
         self.token = token
         self.kwargs = kwargs
         super_kwargs = {k: kwargs.pop(k)
                         for k in ['use_listings_cache', 'listings_expiry_time', 'max_paths']
                         if k in kwargs}  # passed to fsspec superclass
+        super().__init__(loop=loop, asynchronous=asynchronous, **super_kwargs)
 
-        if client_kwargs is None:
-            client_kwargs = {}
-        if config_kwargs is None:
-            config_kwargs = {}
         self.default_block_size = default_block_size or self.default_block_size
         self.default_fill_cache = default_fill_cache
         self.default_cache_type = default_cache_type
         self.version_aware = version_aware
-        self.client_kwargs = client_kwargs
-        self.config_kwargs = config_kwargs
+        self.client_kwargs = client_kwargs or {}
+        self.config_kwargs = config_kwargs or {}
         self.req_kw = {'RequestPayer': 'requester'} if requester_pays else {}
         self.s3_additional_kwargs = s3_additional_kwargs or {}
         self.use_ssl = use_ssl
-        self.s3 = self.connect()
+        if not asynchronous:
+            self.connect()
+            weakref.finalize(self, sync, self.loop, self._s3.close)
+        else:
+            self._s3
         self._kwargs_helper = ParamKwargsHelper(self.s3)
-        super().__init__(**super_kwargs)
+
+    @property
+    def s3(self):
+        if self._s3 is None:
+            raise RuntimeError("please await ``._connect`` before anything else")
+        return self._s3
 
     def _filter_kwargs(self, s3_method, kwargs):
         return self._kwargs_helper.filter_dict(s3_method.__name__, kwargs)
 
-    def _call_s3(self, method, *akwarglist, **kwargs):
+    async def _call_s3(self, method, *akwarglist, **kwargs):
         kw2 = kwargs.copy()
         kw2.pop('Body', None)
         logger.debug("CALL: %s - %s - %s" % (method.__name__, akwarglist, kw2))
         additional_kwargs = self._get_s3_method_kwargs(method, *akwarglist,
                                                        **kwargs)
-        return method(**additional_kwargs)
+        for i in range(self.retries):
+            try:
+                return await method(**additional_kwargs)
+            except S3_RETRYABLE_ERRORS as e:
+                err - e
+                time.sleep(1.7**i * 0.1)
```

Comment:
should this be an async sleep?

Reply:
additionally i'd also put something like `min(15, 1.7**i * 0.1)` just to avoid the sleeping time exploding in weird cases

Reply:
Good poin!